src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -0,0 +1,559 @@
+        int earliestEndTimeOfFrequencyEntry = schedule.endTimes[frequencyEntryIdx] - headway + travelTimeFromStartOfTrip;
+
+        if (earliestEndTimeOfFrequencyEntry < earliestTime) return -1;
+
+        // board pessimistically assuming the entry is already running
+        int latestBoardTimeAssumingEntryIsAlreadyRunning = earliestTime + headway;
+        // figure out the latest departure time of this trip at this stop
+        int latestBoardTimeOfFirstTrip = schedule.startTimes[frequencyEntryIdx] + headway + travelTimeFromStartOfTrip;
+        // return the max of those two
+        return Math.max(latestBoardTimeAssumingEntryIsAlreadyRunning, latestBoardTimeOfFirstTrip);
+    }
+
+    private void doTransfers (RaptorState state) {
+        // avoid integer casts in tight loop below
+        int walkSpeedMillimetersPerSecond = (int) (request.walkSpeed * 1000);
+        int maxWalkMillimeters = (int) (request.walkSpeed * request.maxWalkTime * 60 * 1000);
+
+        for (int stop = state.nonTransferStopsTouched.nextSetBit(0); stop > -1; stop = state.nonTransferStopsTouched.nextSetBit(stop + 1)) {
+            // no need to consider loop transfers, since we don't mark patterns here any more
+            // loop transfers are already included by virtue of those stops having been reached
+            TIntList transfersFromStop = transit.transfersForStop.get(stop);
+            if (transfersFromStop != null) {
+                for (int stopIdx = 0; stopIdx < transfersFromStop.size(); stopIdx += 2) {
+                    int targetStop = transfersFromStop.get(stopIdx);
+                    int distanceToTargetStopMillimeters = transfersFromStop.get(stopIdx + 1);
+
+                    if (distanceToTargetStopMillimeters < maxWalkMillimeters) {
+                        // transfer length to stop is acceptable
+                        int walkTimeToTargetStopSeconds = distanceToTargetStopMillimeters / walkSpeedMillimetersPerSecond;
+                        int timeAtTargetStop = state.bestNonTransferTimes[stop] + walkTimeToTargetStopSeconds;
+
+                        if (walkTimeToTargetStopSeconds < 0) {
+                            LOG.error("Negative transfer time!!");
+                        }
+
+                        state.setTimeAtStop(targetStop, timeAtTargetStop, -1, stop, 0, 0, true);
+                    }
+                }
+            }
+        }
+    }
+
+    /** Get a list of the internal IDs of the patterns touched using the given index (frequency or scheduled) */
+    private BitSet getPatternsTouchedForStops(RaptorState state, int[] index) {
+        BitSet patternsTouched = new BitSet();
+
+        for (int stop = state.bestStopsTouched.nextSetBit(0); stop >= 0; stop = state.bestStopsTouched.nextSetBit(stop + 1)) {
+            final int finalStop = stop;
+            transit.patternsForStop.get(stop).forEach(originalPattern -> {
+                int filteredPattern = index[originalPattern];
+
```

Comment:
We should be able to `return true` here immediately after seeing filteredPattern < 0, simplifying the conditional below, avoiding the sourcePatternIndex computation, and making it clearer why the iteration is being skipped over.